### ISSUE TITLE
False negative passing null to j.u.function.*

### ISSUE THREAD

**Issue by commonquail** [uber/NullAway#1437]
I imagine this is a well known problem but I was unable to tie it directly to one of the several other issues about generics, lambdas, and method references.

When I pass null to seemingly any of the `java.util.function` `@FunctionalInterface` instances, where the type parameter is `@NonNull`, NullAway does not flag that violation. More surprisingly, NullAway flags equivalent violations in identical local definitions. Demonstration:

```java
import java.util.function.Function;
import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

@NullMarked
class WitherBug {
  void baseline(Function<Object, String> f) {
    // true positive
    nonnull(null);
    // true negative
    f.apply(new Object());
  }

  void g1(
      Function<@Nullable Object, String> f1,
      Function<@NonNull Object, String> f2,
      Function<Object, String> f3) {
    // true negative
    f1.apply(null);
    // FALSE NEGATIVE
    f2.apply(null);
    f3.apply(null);
  }

  // This is fine
  void g2(F1 f1, F2 f2, F3 f3, Func<Object, String> f4) {
    // true negative
    f1.apply(null);
    // true positive
    f2.apply(null);
    f3.apply(null);
    f4.apply(null);
  }

  void nonnull(Object o) { }

  @FunctionalInterface
  interface F1 {
    String apply(@Nullable Object o);
  }

  @FunctionalInterface
  interface F2 {
    String apply(Object o);
  }

  @FunctionalInterface
  interface F3 {
    String apply(Object o);
  }

  @FunctionalInterface
  interface Func<T, R> {
    R apply(T t);
  }
}
```

The behaviour seems to be unaffected by `OnlyNullMarked`, `AnnotatedPackages`, and `JSpecifyMode`, as well as the placement of (a containing) `@NullMarked`. `@FunctionalInterface` is irrelevant, too, and only included for parity.

Built with

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>example</groupId>
  <artifactId>playground</artifactId>
  <version>1.0-SNAPSHOT</version>

  <properties>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <maven.compiler.release>21</maven.compiler.release>
  </properties>

  <dependencies>
    <dependency>
      <groupId>org.jspecify</groupId>
      <artifactId>jspecify</artifactId>
      <version>1.0.0</version>
    </dependency>
  </dependencies>

  <build>
    <pluginManagement>
      <plugins>
        <plugin>
          <artifactId>maven-compiler-plugin</artifactId>
          <version>3.14.1</version>
          <configuration>
            <fork>true</fork>
            <compilerArgs>
              <arg>-XDcompilePolicy=simple</arg>
              <arg>-XDaddTypeAnnotationsToSymbol=true</arg>
              <arg>--should-stop=ifError=FLOW</arg>
              <arg>-Xplugin:ErrorProne -XepDisableAllChecks \
                  -Xep:NullAway:WARN \
                  -XepOpt:NullAway:OnlyNullMarked=true \
                  -XepOpt:NullAway:WarnOnGenericInferenceFailure:true \
                  -XepOpt:NullAway:JSpecifyMode:true</arg>
              <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED</arg>
              <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED</arg>
              <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED</arg>
              <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED</arg>
              <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED</arg>
              <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED</arg>
              <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED</arg>
              <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED</arg>
              <arg>-J--add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED</arg>
              <arg>-J--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED</arg>
            </compilerArgs>
            <annotationProcessorPaths>
              <path>
                <groupId>com.google.errorprone</groupId>
                <artifactId>error_prone_core</artifactId>
                <version>2.46.0</version>
              </path>
              <path>
                <groupId>com.uber.nullaway</groupId>
                <artifactId>nullaway</artifactId>
                <version>0.12.15</version>
              </path>
            </annotationProcessorPaths>
          </configuration>
        </plugin>
      </plugins>
    </pluginManagement>
  </build>
</project>
```

**Comment by commonquail** [comment]
Here's something: https://github.com/commonquail/nullaway-1437

Gradle works (although only with `JSpecifyMode=true`...?), Maven never works. Do they end up compiling differently somehow?

**Comment by msridhar** [comment]
Thanks for the example.  I figured out the issue.  On [this line](https://github.com/commonquail/nullaway-1437/blob/e42c53ef8cfae4acc53b747e5f68d0427ea5bfb5/pom.xml#L38) it should read `-XepOpt:NullAway:JSpecifyMode=true` (`=` instead of the last `:`).  Unfortunately there doesn't seem to be validation to catch the syntax error...sorry about it.

And, yes, this only works with JSpecifyMode on, since that enables the generics checking, which NullAway did not support before.

**Comment by msridhar** [comment]
As a small follow up, you can see [here](https://github.com/google/error-prone/blob/4fdce105d9c5d9948c086f2112553421b142a2fa/check_api/src/test/java/com/google/errorprone/ErrorProneFlagsTest.java#L33-L44) that there can be colons inside flag names for Error Prone, so the previous way the command-line arg was written led to a flag with name `JSpecifyMode:true` being created
